File: src/main/java/org/vors/pairbot/service/NotificationService.java
package org.vors.pairbot.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
import org.vors.pairbot.model.Event;
import org.vors.pairbot.model.Participant;
import org.vors.pairbot.model.UserInfo;
import org.vors.pairbot.repository.UserRepository;

import javax.transaction.Transactional;
import java.time.Duration;
import java.util.List;

@Transactional
@Component
public class NotificationService {
    private Logger LOG = LoggerFactory.getLogger(getClass());

    @Autowired
    private UserService userService;
    @Autowired
    private ChatService chatService;
    @Autowired
    private MessageService messageService;
    @Autowired
    private UserRepository userRepository;

    public void notifyAllUpcomingIn(Duration expiringIn, Duration scanPeriod) {
        LOG.info("Start notify about upcoming events");
        List<Event> events = userService.findUpcomingEvents(expiringIn, scanPeriod);
        LOG.info("Found {} people to notify", events.size());

        events.forEach(this::notifyUpcoming);

        LOG.info("End notify about upcoming events");
    }

    private void notifyUpcoming(Event event) {
        try {
            event.getParticipants().forEach(this::notifyUpcoming);
        } catch (Exception e) {
            LOG.error("Can't notify about upcoming event pk = {}\n{}", event.getPk(), e.toString(), e);
        }
    }

    private void notifyUpcoming(Participant participant) {
        UserInfo user = participant.getUser();
        Integer userId = user.getUserId();
        LOG.info("Notifying upcoming user {}", userId);
        try {
            Integer sentMessage = messageService.sendMessage(
                    messageService.getUpcomingNotificationMessage(participant));

            LOG.info("Notified upcoming user {}, id {}", user.getFirstName(), userId);

            user.setLastMessageId(sentMessage);

            userRepository.save(user);
        } catch (TelegramApiException e) {
            LOG.error("Can't notify expiring user {}\n{}", user.getPk(), e.toString(), e);
        }
    }

}

File: src/main/java/org/vors/pairbot/service/NotificationService.kt
package org.vors.pairbot.service

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.telegram.telegrambots.meta.exceptions.TelegramApiException
import org.vors.pairbot.model.Event
import org.vors.pairbot.model.Participant
import org.vors.pairbot.repository.UserRepository
import java.time.Duration
import javax.transaction.Transactional

@Transactional
@Component
open class NotificationService(
        open var userService: UserService,
        open var chatService: ChatService,
        open var messageService: MessageService,
        open var userRepository: UserRepository
) {
    open var log = LoggerFactory.getLogger(javaClass)

    fun notifyAllUpcomingIn(expiringIn: Duration, scanPeriod: Duration) {
        log.info("Start notify about upcoming events")
        val events = userService.findUpcomingEvents(expiringIn, scanPeriod)
        log.info("Found {} people to notify", events.size)

        events.forEach { this.notifyUpcoming(it) }

        log.info("End notify about upcoming events")
    }

    private fun notifyUpcoming(event: Event) {
        try {
            event.participants.forEach { this.notifyUpcoming(it) }
        } catch (e: Exception) {
            log.error("Can't notify about upcoming event pk = {}\n{}", event.pk, e.toString(), e)
        }

    }

    private fun notifyUpcoming(participant: Participant) {
        val user = participant.user
        val userId = user.userId
        log.info("Notifying upcoming user {}", userId)
        try {
            val sentMessage = messageService.sendMessage(
                    messageService.getUpcomingNotificationMessage(participant))

            log.info("Notified upcoming user {}, id {}", user.firstName, userId)

            user.lastMessageId = sentMessage

            userRepository.save(user)
        } catch (e: TelegramApiException) {
            log.error("Can't notify expiring user {}\n{}", user.pk, e.toString(), e)
        }

    }

}
